Shani/branch fixed

shani-arnon opened the pull request:
| Q                | A   |
| ---------------- | --- |
| Bug fix?         | ✔/✖ |
| New feature?     | ✖ |
| Breaking change? | ✖ |
| Deprecations?    | ✖ |
| Documentation?   | ✖ |
| Tests added?     | ✖ |
| Types added?     | ✖ |


- New branch includes two issues (added cache + expand suite)

dependabot-preview[bot] committed:
build(deps-dev): bump eslint-config-prettier from 7.1.0 to 7.2.0

Bumps [eslint-config-prettier](https://github.com/prettier/eslint-config-prettier) from 7.1.0 to 7.2.0.
- [Release notes](https://github.com/prettier/eslint-config-prettier/releases)
- [Changelog](https://github.com/prettier/eslint-config-prettier/blob/main/CHANGELOG.md)
- [Commits](https://github.com/prettier/eslint-config-prettier/compare/v7.1.0...v7.2.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump fs-extra from 9.0.1 to 9.1.0

Bumps [fs-extra](https://github.com/jprichardson/node-fs-extra) from 9.0.1 to 9.1.0.
- [Release notes](https://github.com/jprichardson/node-fs-extra/releases)
- [Changelog](https://github.com/jprichardson/node-fs-extra/blob/master/CHANGELOG.md)
- [Commits](https://github.com/jprichardson/node-fs-extra/compare/9.0.1...9.1.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump rollup from 2.35.1 to 2.38.2

Bumps [rollup](https://github.com/rollup/rollup) from 2.35.1 to 2.38.2.
- [Release notes](https://github.com/rollup/rollup/releases)
- [Changelog](https://github.com/rollup/rollup/blob/master/CHANGELOG.md)
- [Commits](https://github.com/rollup/rollup/compare/v2.35.1...v2.38.2)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump @rollup/plugin-commonjs from 17.0.0 to 17.1.0

Bumps [@rollup/plugin-commonjs](https://github.com/rollup/plugins) from 17.0.0 to 17.1.0.
- [Release notes](https://github.com/rollup/plugins/releases)
- [Commits](https://github.com/rollup/plugins/compare/commonjs-v17.0.0...commonjs-v17.1.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump husky from 4.3.6 to 4.3.8

Bumps [husky](https://github.com/typicode/husky) from 4.3.6 to 4.3.8.
- [Release notes](https://github.com/typicode/husky/releases)
- [Commits](https://github.com/typicode/husky/compare/v4.3.6...v4.3.8)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump @rollup/plugin-babel from 5.2.2 to 5.2.3

Bumps [@rollup/plugin-babel](https://github.com/rollup/plugins) from 5.2.2 to 5.2.3.
- [Release notes](https://github.com/rollup/plugins/releases)
- [Commits](https://github.com/rollup/plugins/compare/babel-v5.2.2...babel-v5.2.3)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump faker from 5.1.0 to 5.2.0

Bumps [faker](https://github.com/Marak/Faker.js) from 5.1.0 to 5.2.0.
- [Release notes](https://github.com/Marak/Faker.js/releases)
- [Changelog](https://github.com/Marak/faker.js/blob/master/CHANGELOG.md)
- [Commits](https://github.com/Marak/Faker.js/compare/v5.1.0...v5.2.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump @rollup/plugin-alias from 3.1.1 to 3.1.2

Bumps [@rollup/plugin-alias](https://github.com/rollup/plugins) from 3.1.1 to 3.1.2.
- [Release notes](https://github.com/rollup/plugins/releases)
- [Commits](https://github.com/rollup/plugins/compare/alias-v3.1.1...alias-v3.1.2)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump eslint from 7.16.0 to 7.19.0

Bumps [eslint](https://github.com/eslint/eslint) from 7.16.0 to 7.19.0.
- [Release notes](https://github.com/eslint/eslint/releases)
- [Changelog](https://github.com/eslint/eslint/blob/master/CHANGELOG.md)
- [Commits](https://github.com/eslint/eslint/compare/v7.16.0...v7.19.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump @rollup/plugin-node-resolve from 11.0.1 to 11.1.1

Bumps [@rollup/plugin-node-resolve](https://github.com/rollup/plugins) from 11.0.1 to 11.1.1.
- [Release notes](https://github.com/rollup/plugins/releases)
- [Commits](https://github.com/rollup/plugins/compare/commonjs-v11.0.1...node-resolve-v11.1.1)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

yakkomajuri committed:
patch(vest): add "remove" type, fix typo (#578)

Commit:
Good first issue - expand suite methods

Commit:
expend interface and suite class get error methods

Commit:
linter error fixed

Commit:
add cashe around collectFailureMessages

Commit:
fixed linter

dependabot-preview[bot] committed:
build(deps-dev): bump lodash from 4.17.20 to 4.17.21

Bumps [lodash](https://github.com/lodash/lodash) from 4.17.20 to 4.17.21.
- [Release notes](https://github.com/lodash/lodash/releases)
- [Commits](https://github.com/lodash/lodash/compare/4.17.20...4.17.21)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump @babel/core from 7.12.10 to 7.13.8

Bumps [@babel/core](https://github.com/babel/babel/tree/HEAD/packages/babel-core) from 7.12.10 to 7.13.8.
- [Release notes](https://github.com/babel/babel/releases)
- [Changelog](https://github.com/babel/babel/blob/main/CHANGELOG.md)
- [Commits](https://github.com/babel/babel/commits/v7.13.8/packages/babel-core)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump @rollup/plugin-node-resolve from 11.1.1 to 11.2.0

Bumps [@rollup/plugin-node-resolve](https://github.com/rollup/plugins) from 11.1.1 to 11.2.0.
- [Release notes](https://github.com/rollup/plugins/releases)
- [Commits](https://github.com/rollup/plugins/compare/node-resolve-v11.1.1...node-resolve-v11.2.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump date-fns from 2.16.1 to 2.17.0

Bumps [date-fns](https://github.com/date-fns/date-fns) from 2.16.1 to 2.17.0.
- [Release notes](https://github.com/date-fns/date-fns/releases)
- [Changelog](https://github.com/date-fns/date-fns/blob/master/CHANGELOG.md)
- [Commits](https://github.com/date-fns/date-fns/compare/v2.16.1...v2.17.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump @babel/register from 7.12.10 to 7.13.8

Bumps [@babel/register](https://github.com/babel/babel/tree/HEAD/packages/babel-register) from 7.12.10 to 7.13.8.
- [Release notes](https://github.com/babel/babel/releases)
- [Changelog](https://github.com/babel/babel/blob/main/CHANGELOG.md)
- [Commits](https://github.com/babel/babel/commits/v7.13.8/packages/babel-register)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump eslint-config-prettier from 7.2.0 to 8.1.0

Bumps [eslint-config-prettier](https://github.com/prettier/eslint-config-prettier) from 7.2.0 to 8.1.0.
- [Release notes](https://github.com/prettier/eslint-config-prettier/releases)
- [Changelog](https://github.com/prettier/eslint-config-prettier/blob/main/CHANGELOG.md)
- [Commits](https://github.com/prettier/eslint-config-prettier/compare/v7.2.0...v8.1.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump @rollup/plugin-replace from 2.3.4 to 2.4.1

Bumps [@rollup/plugin-replace](https://github.com/rollup/plugins) from 2.3.4 to 2.4.1.
- [Release notes](https://github.com/rollup/plugins/releases)
- [Commits](https://github.com/rollup/plugins/compare/replace-v2.3.4...replace-v2.4.1)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump typescript from 4.1.3 to 4.2.2

Bumps [typescript](https://github.com/Microsoft/TypeScript) from 4.1.3 to 4.2.2.
- [Release notes](https://github.com/Microsoft/TypeScript/releases)
- [Commits](https://github.com/Microsoft/TypeScript/compare/v4.1.3...v4.2.2)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump husky from 4.3.8 to 5.1.2

Bumps [husky](https://github.com/typicode/husky) from 4.3.8 to 5.1.2.
- [Release notes](https://github.com/typicode/husky/releases)
- [Commits](https://github.com/typicode/husky/compare/v4.3.8...v5.1.2)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump rollup from 2.38.2 to 2.40.0

Bumps [rollup](https://github.com/rollup/rollup) from 2.38.2 to 2.40.0.
- [Release notes](https://github.com/rollup/rollup/releases)
- [Changelog](https://github.com/rollup/rollup/blob/master/CHANGELOG.md)
- [Commits](https://github.com/rollup/rollup/compare/v2.38.2...v2.40.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump @rollup/plugin-babel from 5.2.3 to 5.3.0

Bumps [@rollup/plugin-babel](https://github.com/rollup/plugins) from 5.2.3 to 5.3.0.
- [Release notes](https://github.com/rollup/plugins/releases)
- [Commits](https://github.com/rollup/plugins/compare/babel-v5.2.3...babel-v5.3.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump @babel/plugin-transform-object-assign

Bumps [@babel/plugin-transform-object-assign](https://github.com/babel/babel/tree/HEAD/packages/babel-plugin-transform-object-assign) from 7.12.1 to 7.12.13.
- [Release notes](https://github.com/babel/babel/releases)
- [Changelog](https://github.com/babel/babel/blob/main/CHANGELOG.md)
- [Commits](https://github.com/babel/babel/commits/v7.12.13/packages/babel-plugin-transform-object-assign)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump @babel/cli from 7.12.10 to 7.13.0

Bumps [@babel/cli](https://github.com/babel/babel/tree/HEAD/packages/babel-cli) from 7.12.10 to 7.13.0.
- [Release notes](https://github.com/babel/babel/releases)
- [Changelog](https://github.com/babel/babel/blob/main/CHANGELOG.md)
- [Commits](https://github.com/babel/babel/commits/v7.13.0/packages/babel-cli)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump @babel/plugin-proposal-optional-chaining

Bumps [@babel/plugin-proposal-optional-chaining](https://github.com/babel/babel/tree/HEAD/packages/babel-plugin-proposal-optional-chaining) from 7.12.7 to 7.13.8.
- [Release notes](https://github.com/babel/babel/releases)
- [Changelog](https://github.com/babel/babel/blob/main/CHANGELOG.md)
- [Commits](https://github.com/babel/babel/commits/v7.13.8/packages/babel-plugin-proposal-optional-chaining)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump @babel/plugin-transform-runtime

Bumps [@babel/plugin-transform-runtime](https://github.com/babel/babel/tree/HEAD/packages/babel-plugin-transform-runtime) from 7.12.10 to 7.13.8.
- [Release notes](https://github.com/babel/babel/releases)
- [Changelog](https://github.com/babel/babel/blob/main/CHANGELOG.md)
- [Commits](https://github.com/babel/babel/commits/v7.13.8/packages/babel-plugin-transform-runtime)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps-dev): bump @babel/preset-env from 7.12.11 to 7.13.8

Bumps [@babel/preset-env](https://github.com/babel/babel/tree/HEAD/packages/babel-preset-env) from 7.12.11 to 7.13.8.
- [Release notes](https://github.com/babel/babel/releases)
- [Changelog](https://github.com/babel/babel/blob/main/CHANGELOG.md)
- [Commits](https://github.com/babel/babel/commits/v7.13.8/packages/babel-preset-env)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

ealush committed:
patch: remove package-lock.json

ealush commented on packages/vest/src/core/produce/hasFailuresByGroup.js:
```diff
@@ -1,4 +1,4 @@
-import { hasLogic } from 'hasFaillures';
+import { hasLogic } from 'hasFailures';
```
🤦🏻‍♂️

shani-arnon replied:
🤓

ealush commented on packages/vest/src/core/produce/collectFailureMessages.js:
```diff
@@ -11,35 +15,28 @@ import useTestObjects from 'useTestObjects';
+  const ctxRef = { stateRef };
+
+  return cache(
+    [testObjects, severity, options], context.bind(ctxRef, () => {
+      const res = testObjects.reduce((collector, testObject) => {
+        if (group && testObject.groupName !== group) return collector;
+        if (fieldName && testObject.fieldName !== fieldName) return collector;
+        if (!testObject.failed) return collector;
+        if (isMatchingSeverityProfile(severity, testObject)) return collector;
+
+        collector[testObject.fieldName] = (
+          collector[testObject.fieldName] || []
+        ).concat(testObject.statement);
+
+        return collector;
+      }, {});
+
+      if (fieldName) return res[fieldName] || [];
+      return res;
+    })
```
Are you sure we need to bind the state to the context here? From the looks of it we do not use the state in this function or in any of the functions being called by it. I may be wrong about this - let's have a quick run of the tests without it to see if it actually works as expected.

```suggestion
    [testObjects, severity, options], () => {
      const res = testObjects.reduce((collector, testObject) => {
        if (group && testObject.groupName !== group) return collector;
        if (fieldName && testObject.fieldName !== fieldName) return collector;
        if (!testObject.failed) return collector;
        if (isMatchingSeverityProfile(severity, testObject)) return collector;

        collector[testObject.fieldName] = (
          collector[testObject.fieldName] || []
        ).concat(testObject.statement);

        return collector;
      }, {});

      if (fieldName) return res[fieldName] || [];
      return res;
    }
```

shani-arnon replied:
removed state bind. ran tests and works as expected

ealush commented on packages/vest/src/core/suite/createSuite.js:
```diff
@@ -67,6 +67,14 @@ const createSuite = withArgs(args => {
       }
     });
   });
+  suite.hasErrors = (fieldName) => suite.get().hasErrors(fieldName);
+  suite.hasWarnings = (fieldName) => suite.get().hasWarnings(fieldName);
+  suite.getErrors = (fieldName) => suite.get().getErrors(fieldName);
+  suite.getWarnings = (fieldName) => suite.get().getWarnings(fieldName);
+  suite.hasErrorsByGroup = (groupName, fieldName) => suite.get().hasErrorsByGroup(groupName, fieldName);
+  suite.hasWarningsByGroup = (groupName, fieldName) => suite.get().hasWarningsByGroup(groupName, fieldName);
+  suite.getErrorsByGroup = (groupName, fieldName) => suite.get().getErrorsByGroup(groupName, fieldName);
+  suite.getWarningsByGroup = (groupName, fieldName) => suite.get().getWarningsByGroup(groupName, fieldName);
```
Nice! This looks pretty good. Let's also add a couple of tests for these added functions. Nothing big, we only need to check that they match the original implementation.

Also, do you want to handle the documentation as well or should I?

ealush commented on packages/vest/src/typings/vest.d.ts:
```diff
@@ -167,6 +167,27 @@ interface ICreateResult {
   get: () => DraftResult;
   reset: () => void;
   remove: (fieldName: string) => void;
+  
+  hasErrors: (fieldName?: string) => boolean;
+  hasWarnings: (fieldName?: string) => boolean;
+  
+  getErrors(): { [fieldName: string]: string[] };
+  getErrors(fieldName: string): string[];
+
+  getWarnings(): { [fieldName: string]: string[] };
+  getWarnings(fieldName: string): string[];
+  
+  hasErrorsByGroup(groupName: string): boolean;
+  hasErrorsByGroup(groupName: string, fieldName: string): boolean;
+  
+  hasWarningsByGroup(groupName: string): boolean;
+  hasWarningsByGroup(groupName: string, fieldName: string): boolean;
+  
+  getErrorsByGroup(groupName: string): { [fieldName: string]: string[] };
+  getErrorsByGroup(groupName: string, fieldName: string): string[];
+  
+  getWarningsByGroup(groupName: string): { [fieldName: string]: string[] };
+  getWarningsByGroup(groupName: string, fieldName: string): string[];
```
Nice!
Here's an enhancement idea: Since all those are already typed here https://github.com/ealush/vest/blob/latest/packages/vest/src/typings/vestResult.d.ts#L28-L85 why not move them to their own type and use in both places? Would make maintenance way easier

shani-arnon replied:
Done.

ealush reviewed (commented):
Great job @shani-arnon! I left a couple of comments but overall this is looking pretty good.

Let's take care of these little details and get this one merged ✌️

Thank you for making Vest better 🙏🏻

Commit:
remove unused state bind

Commit:
remove function from vest.d.ts to vestResult

Commit:
add tests just for making sure nothing was broken

ealush committed:
minor(vest): add skipped tests to result as well (#600)

galitpauz committed:
docs: Update getting-started vest script tag version number (#603)

codechips committed:
docs: Add example for Svelte (#607)

Co-authored-by: Evyatar <code@ealush.com>

Commit:
added new test for suite methods - sanity

ronen-e committed:
feat(vest): add state subscription (#606)

dependabot-preview[bot] committed:
build(deps): [security] bump y18n from 4.0.0 to 4.0.1

Bumps [y18n](https://github.com/yargs/y18n) from 4.0.0 to 4.0.1. **This update includes a security fix.**
- [Release notes](https://github.com/yargs/y18n/releases)
- [Changelog](https://github.com/yargs/y18n/blob/master/CHANGELOG.md)
- [Commits](https://github.com/yargs/y18n/commits)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

ealush committed:
patch(vest): make state independent of context (#609)

ealush committed:
patch(vest): use vast state library

ealush committed:
patch(vest): upgrade context package

Commit:
fixed conflicts

Commit:
--

Commit:
conflicts try

Commit:
Merge branch 'ealush_latest' into shani/branch-fixed
